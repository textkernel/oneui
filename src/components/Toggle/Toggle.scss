@import '../../themes/oneui/oneui.scss';

.Toggle {
    display: flex;
    align-items: center;
    cursor: pointer;

    &__toggleContainer {
        width: 36px;
        height: 20px;
        border-radius: 20px;
        position: relative;
        background-color: var(--color-neutral);
        transition: border-color 300ms;
        flex-shrink: 0;

        .Toggle__realCheckbox {
            position: absolute;
            clip: rect(0,0,0,0);

            & + .Toggle__toggleButton {
                position: absolute;
                top: 0;
                left: 0;
                right: 0;
                bottom: 0;
                border-radius: 20px;
                transition: all 300ms;

                &::before {
                    content: '';
                    display: inline-block;
                    width: 18px;
                    height: 18px;
                    margin: 1px;
                    background-color: white;
                    border-radius: 50%;
                    transition: all 300ms ease-in-out;
                }

            }

            &:checked {
                + .Toggle__toggleButton {
                    background-color: var(--color-good-40);

                    &::before{
                        margin-left: 17px;
                    }
                }

                &[disabled] + .Toggle__toggleButton {
                    background-color: var(--color-good-25);
                }
            }
        }
    }

    &__text {
        padding: var(--spacing-normal) 0 var(--spacing-normal) var(--spacing-2x);
    }

    &--disabled {
        cursor: not-allowed;
    }
}

