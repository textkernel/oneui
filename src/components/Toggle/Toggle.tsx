import * as React from 'react';
import { bem } from '../../utils';
import { Text } from '../Text';
import styles from './Toggle.scss';

export interface Props extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'type'> {
    /** A unique id to reference this toggle */
    id: string;
    /** Label related to toggle */
    children?: string;
    /** Should be Toggle disabled or not */
    disabled?: boolean;
}

const { block, elem } = bem('Toggle', styles);

export const Toggle = React.forwardRef<HTMLInputElement, Props>((props, ref) => {
    const { children, disabled, id, ...rest } = props;
    return (
        <label {...block(props)} htmlFor={id}>
            <div {...elem('toggleContainer')}>
                <input
                    {...elem('realCheckbox')}
                    {...rest}
                    ref={ref}
                    type="checkbox"
                    id={id}
                    disabled={disabled}
                />
                <div {...elem('toggleButton')} />
            </div>
            {children && (
                <Text {...elem('text')} inline context={disabled ? 'muted' : 'default'}>
                    {children}
                </Text>
            )}
        </label>
    );
});

Toggle.displayName = 'Toggle';

Toggle.defaultProps = {
    disabled: false,
};
