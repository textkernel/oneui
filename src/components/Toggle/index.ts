export { Toggle, Props as ToggleProps } from './Toggle';
