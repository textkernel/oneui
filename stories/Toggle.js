import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, text, withKnobs } from '@storybook/addon-knobs';
import { Toggle } from '@textkernel/oneui';

storiesOf('Molecules/Toggle', module)
    .addDecorator(withKnobs)
    .add('Toggle with label', () => (
        <Toggle
            disabled={boolean('disabled', false)}
            id={text('id', 'my-toggle-1')}
            onChange={(event) => {
                console.log('Toggle state changed. event.target.checked: ', event.target.checked);
            }}
        >
            {text('children', 'This is title!')}
        </Toggle>
    ))
    .add('Toggle without label', () => (
        <Toggle
            disabled={boolean('disabled', false)}
            id={text('id', 'my-toggle-2')}
            onChange={(event) => {
                console.log('Toggle state changed. event.target.checked: ', event.target.checked);
            }}
        />
    ));
