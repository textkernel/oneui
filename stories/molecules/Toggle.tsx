import * as React from 'react';
import { Toggle } from '@textkernel/oneui';

export default {
    title: 'Molecules/Toggle',
    component: Toggle,
};

export const ToggleWithLabel = (args) => <Toggle {...args} />;
ToggleWithLabel.storyName = 'Toggle with label';
ToggleWithLabel.args = {
    id: 'my-toggle-1',
    children: 'This is title!',
};

export const ToggleWithoutLabel = (args) => <Toggle {...args} />;
ToggleWithoutLabel.storyName = 'Toggle without label';
ToggleWithoutLabel.args = {
    id: 'my-toggle-2',
};
