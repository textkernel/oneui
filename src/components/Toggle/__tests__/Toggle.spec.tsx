import React from 'react';
import { render, RenderResult, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Toggle } from '../Toggle';

describe('<Toggle> that renders a toggle', () => {
    let view: RenderResult;

    const getCheckboxByName = (inputName) => {
        return screen.getByRole('checkbox', { name: `${inputName}` });
    };

    it('should render Toggle correctly with label', () => {
        view = render(<Toggle id="1">Title is here</Toggle>);

        expect(view.container).toMatchSnapshot();
        expect(getCheckboxByName('Title is here')).toBeVisible();
    });

    it('should render Toggle correctly without label', () => {
        view = render(<Toggle id="1-2" />);

        expect(view.container).toMatchSnapshot();
        expect(getCheckboxByName('')).toBeVisible();
    });

    it('should rendered disabled toggle correctly', () => {
        view = render(
            <Toggle id="4" disabled>
                Cannot do anything =(
            </Toggle>
        );

        const checkbox = getCheckboxByName('Cannot do anything =(');

        expect(view.container).toMatchSnapshot();
        expect(checkbox).toBeVisible();
        expect(checkbox).toHaveAttribute('disabled');
    });
});
