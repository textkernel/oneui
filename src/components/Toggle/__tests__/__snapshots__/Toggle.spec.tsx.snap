// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle> that renders a toggle should render Toggle correctly with label 1`] = `
<div>
  <label
    class="Toggle"
    for="1"
  >
    <div
      class="Toggle__toggleContainer"
    >
      <input
        class="Toggle__realCheckbox"
        id="1"
        type="checkbox"
      />
      <div
        class="Toggle__toggleButton"
      />
    </div>
    <span
      class="Toggle__text"
    >
      Title is here
    </span>
  </label>
</div>
`;

exports[`<Toggle> that renders a toggle should render Toggle correctly without label 1`] = `
<div>
  <label
    class="Toggle"
    for="1-2"
  >
    <div
      class="Toggle__toggleContainer"
    >
      <input
        class="Toggle__realCheckbox"
        id="1-2"
        type="checkbox"
      />
      <div
        class="Toggle__toggleButton"
      />
    </div>
  </label>
</div>
`;

exports[`<Toggle> that renders a toggle should rendered disabled toggle correctly 1`] = `
<div>
  <label
    class="Toggle Toggle--disabled"
    for="4"
  >
    <div
      class="Toggle__toggleContainer"
    >
      <input
        class="Toggle__realCheckbox"
        disabled=""
        id="4"
        type="checkbox"
      />
      <div
        class="Toggle__toggleButton"
      />
    </div>
    <span
      class="Text--context_neutral Toggle__text"
    >
      Cannot do anything =(
    </span>
  </label>
</div>
`;
